webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/queries/LdapQueryProcessor.scala

```diff
@@ -259,22 +262,20 @@ object PostFilterNodeFromInfoService {
     composition match {
       // TODO: there is surely something we can do here
       case Or  =>
-        val foundNodeIds = foundNodeInfos.map(_.id).toSet
-        allNodesInfos.filter { nodeinfo => predicate(nodeinfo, predicates, composition, Some(foundNodeIds)) }
+        foundNodeIds = foundNodeInfos.map(_.id).toSet
+        allNodesInfos.filter { nodeinfo => predicate(nodeinfo, predicates, composition, Some(foundNodeIds)) }.map(_.id)
       case And =>
         if (predicates.isEmpty) {
           // paththru
-          foundNodeInfos
+          foundNodeInfos.map(_.id)
         } else {
-          foundNodeInfos.filter { nodeinfo => predicate(nodeinfo, predicates, composition, None) }
+          foundNodeInfos.flatMap { case nodeinfo => if (predicate(nodeinfo, predicates, composition, None)) { Some(nodeinfo.id) } else None }
```

Comment:
can be directly
```
foundNodeInfos.collect { case nodeinfo if (predicate(nodeinfo, predicates, composition, None)) => nodeinfo.id } 
```

Reply:
actually, it could still be the filter

Reply:
the collect is better, it avoids one creation of collection